### ISSUE TITLE
[Feature Request] PettingZoo support

### ISSUE THREAD

**Issue by MarkHaoxiang** [pytorch/rl#1441]
## Motivation

Add PettingZoo, a multiagent framework, support to TorchRL.

## Solution
Create implementation wrapper.

I would be interested contributing to the library and would like to add PettingZoo support. @vmoens would there be any guidelines / onboarding process for new contributors? Thanks.

**Comment by vmoens** [comment]
Thanks for this suggestion!
Yes this feature would be amazing for TorchRL, we'd love to help you implement that.

Integrating a new wrapper is usually done in the following way:
- You will need a new directory under `torchrl.envs.libs`. We usually write a `Wrapper` class that reads the `step` and `reset` of a nested env and converts it to torchrl format. You can have a look at the others, Gym may be the most relevant in this case?
- For multiagent, the general way we structure data is as this:
  - We use the nesting of specs / data as an indicator of the batch size. Example: you have 3 agents that share one common observation. The observation will be at the root and the agent will be within a nested spec: 
```python
CompositeSpec({
  "shared_obs": DiscreteSpec(shape=()), 
  "agents": CompositeSpec(
    {"single_obs": DiscreteSpec(shape=(n_agents,))},
    shape=(n_agents,)
  ),
  shape=()
)
```
By playing with the key dimensions and shape dimensions of the specs / tensordict we can represent data in a way that is self-explanatory.
  - You can have actions, done and reward placed anywhere you want in the data structure through a CompositeSpec. For instance, you could have `input_spec["_action_spec"] = CompositeSpec({("agent", "action"): DiscreteSpec(...))`. That way you can apply what we talked about in the point above to tell if the action is shared or individual etc.
  - This convention may not align exactly with what PettingZoo is using so some degree of transformation could be needed. Happy to talk about what the best way is to approach this!
  - Regarding testing, a new CI workflow that installs pettingzoo and tests some representative envs is the best way forward. The minimum tests to run are: 
    - testing that the env specs are well defined with `torchrl.envs.utils.check_envs_specs(env)`
    - testing that serial and parallel envs work as well with the same tool
    - testing a single data collection with our data collectors.
  Once you've written the tests I will be happy to help writing the CI workflow!

Anything you need help with or is unclear feel free to reach out!

ccing @matteobettini for context

**Comment by MarkHaoxiang** [comment]
Thanks for the response - I'm currently working on the branch here https://github.com/pytorch/rl/compare/main...MarkHaoxiang:rl:main

My focus is to get all environments training with a parallel environment first, with the construction similar to VMAS but without batching. There may be complications when dealing with heterogeneous spaces, the PettingZoo's AEC model and variable number of agents which could be worth discussion!

**Comment by matteobettini** [comment]
hey @MarkHaoxiang how is the development going? is there a PR you can submit for review?

I am also looking into introducing PettingZoo wrappers in TorchRL in the coming week. so we can coordinate.

Also it would be nice to allow users to choose if to stack agents or keep them in separate dicts (allowing them to use the different philosophies of #1463) with a flag at env construction. I believe as of now your are implementing case 1 if I am not mistaken.

**Comment by MarkHaoxiang** [comment]
Hi @matteobettini

Have been quite busy with other things this week and continuing development slowly. At the moment I have working versions of  Case 1  (multiwalker, simple, pistonball etc) working and in the process of writing tests / utilities. Looking to move to case 2 in the coming few days.

Not sure how well we could implement the AeC model in Torch while avoiding repetitive computation, perhaps along the lines of case 3 with explicit action masking. 

How would you like to collaborate?

**Comment by matteobettini** [comment]
So, regarding the AeC envs, yes the idea would be to use the same interface we use for parallel envs with added action masking. I believe we cannot get rid of this as of now since, being a Tensor-based library, we need to stack data over time.
Case 3 and 2 were introduced to manage heterogeneity amongst agents (with different spaces) but I don't believe that we can do much more than masking for heterogeneity over time (AeC).

Regarding our Petting zoo API, we need a way to group agents. 
For now you have used case one since  (multiwalker, simple, pistonball etc)  contain all homogenous agents, but we need to define our strategy to handle heterogenous agents.

My proposition is this:

## Agents will be grouped by default given their name

Some examples:

in pistonball `agents= ['piston_0', 'piston_1', ..., 'piston_19']` indicates that we will have one group name piston

```
Tensordict(
    "piston":TensorDict({
          "done"
          "action"
          "obs"
          "reward"
     },
      batch_size=[*B,20])
 batch_size=[*B])
)
```

in simple_tag `agents= [adversary_0, adversary_1, adversary_2, agent_0]` indicates that we will have two groups

```
Tensordict(
    "adversary":TensorDict({
          "done"
          "action"
          "obs"
          "reward"
     },
      batch_size=[*B,3])
    "agent":TensorDict({
          "done"
          "action"
          "obs"
          "reward"
     },
      batch_size=[*B])
 batch_size=[*B])
)
```
you get the gist. the agents will be groupes by the str before the _ in their name.

## When groups will be split

We will try to do what outlined above, but it some cases it might not be possible or desired, therefore, groups will be split into separate tensordict entries when:

- **agents within a group have different spaces will lead to that group being split**: an environment with agents  `agents= [agent_0, agent_1, agent_2]` but the agents actually have different spaces. Then the grouping will not happen for that group and we will have

```
TensorDict(
    "agent_0": (
        "obs_a": Tensor,
        "action": Tensor,
        "reward": Tensor,
        "done": Tensor,
    batch_size=[*B]),
     "agent_1": (
        "obs_a": Tensor,
        "action": Tensor,
        "reward": Tensor,
        "done": Tensor,
    batch_size=[*B]),
    "agent_2": (
        "obs_a": Tensor,
        "action": Tensor,
        "reward": Tensor,
        "done": Tensor,
    batch_size=[*B]),
    "state": Tensor,
batch_size=[*B])
```

- **the agent names do not follow the expected convention**: for agents that do not follow the naming convention str_number we will create an group for each. for example  `agents= [agent_0, agent_1, robert]` 
```
TensorDict(
    "agent": (
        "obs_a": Tensor,
        "action": Tensor,
        "reward": Tensor,
        "done": Tensor,
    batch_size=[*B, 2]),
    "robert": (
        "obs_a": Tensor,
        "action": Tensor,
        "reward": Tensor,
        "done": Tensor,
    batch_size=[*B]),
    "state": Tensor,
batch_size=[*B])
```

- **with a flag at construction** . if `PettingZooEnv(group_agents=False)` no stacking of agents will be done and each agent will have a key with its name

What do you think? Does this resonate with you?

**Comment by matteobettini** [comment]
Finally we can allow users to completely define their grouping logic, with

`PettingZooEnv(group_map={“goodies”: [“agent_0”,”agent_1”], “baddies”:[“agent_2”]})`

**Comment by MarkHaoxiang** [comment]
Makes sense! Would it potentially be worth considering nested-grouping logic as well, say (team_0, agent_0) and (team_1, agent_0)?

As for AeC / Parallel, I'll create a PR just for the parallel once that is completed since most environments can be represented as POSGs anyway and it relies on action masking.

**Comment by matteobettini** [comment]
Could you elaborate more in these nested groups? When would they be needed? What is the usecase where normal grouping as in #1463 would not work?

**Comment by matteobettini** [comment]
Hey @MarkHaoxiang, 

For my internship I might need to have this wrapper by the end of this week to run some experiments.

I understand if you do not have the time to ultimate it by then.
I can open my separate PR for the PettingZoo wrapper.

The multiagent conv model you have is cool so defo you can PR that

**Comment by MarkHaoxiang** [comment]
Hi @matteobettini ,

Sorry for the delay. I can open a PR for Parallel Environments and Het agents before Friday, but I am unfamiliar with Action masks and won't likely be able to include that functionality by then.

Let me know if you would prefer to implement the wrapper yourself and I will move the multiagent models into a separate PR!

**Comment by matteobettini** [comment]
Hey, if you don't mind I might take a stab at the wrapper since I am planning to create a few util functions for grouping that we ll reuse across multiagent simulators.

I ll cc you on my PR so that if you want you can review if it aligns with what you thought.

Yes pls open a pr for the model

**Comment by matteobettini** [comment]
PettingZoo is now in main!
